can/zynqmp-can/zynqmp-can.c

```diff
@@ -0,0 +1,552 @@
+	}
+	printf("\n");
+	printf("\t-h             - Print this message\n");
+}
+
+/**
+ * Initializes the clock for the specified CAN peripheral.
+ * Sets the IO_PLL as the clock source and configures the divider.
+ */
+static int init_clock(int can_id)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_devclock;
+
+	/* Set IO_PLL as source clock and set divider:
+	 * IO_PLL / 20 :  1000 MHz / 20 = 50 MHz */
+	ctl.devclock.dev = info[can_id].clk;
+	ctl.devclock.src = 0;
+	ctl.devclock.div0 = 20;
+	ctl.devclock.div1 = 0;
+	ctl.devclock.active = 0x1;
+
+	return platformctl(&ctl);
+}
+
+/**
+ * Configures the MIO pin for CAN RX or TX.
+ * Sets default properties and enables pull-up resistors.
+ */
+static int config_mio_pin(uint32_t pin)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_mio;
+
+	/* Set default properties for pins */
+	ctl.mio.pin = pin;
+	ctl.mio.l0 = ctl.mio.l1 = ctl.mio.l2 = 0;
+	ctl.mio.l3 = 0x1;
+	ctl.mio.config = PCTL_MIO_SLOW_nFAST | PCTL_MIO_PULL_UP_nDOWN | PCTL_MIO_PULL_ENABLE;
+
+	if ((pin == CAN0_RX) || (pin == CAN1_RX)) {
+		ctl.mio.config |= PCTL_MIO_TRI_ENABLE;
+	}
+
+	return platformctl(&ctl);
+}
+
+/**
+ * Resets the specified CAN peripheral using the reset subsystem.
+ */
+static int reset_can_peripheral(int can_id)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_devreset;
+
+	ctl.devreset.dev = info[can_id].rst;
+	ctl.devreset.state = 0;
+
+	return platformctl(&ctl);
+}
+
+/**
+ * CAN interrupt handler.
+ * Handles RX FIFO non-empty, TX FIFO empty, and TX FIFO watermark interrupts.
+ */
+static int can_irq_hdl(unsigned int can_id, void *arg)
+{
+	uint32_t isr = can.base->ISR;
+
+	/* RX FIFO Non-empty IRQ occurred */
+	if (isr & (1 << 7)) {
+		can.base->IER &= ~(1 << 7);
+	}
+
+	/* TX FIFO empty IRQ occurred */
+	if (isr & (1 << 14)) {
+		can.base->IER &= ~(1 << 14);
+	}
+
+	/* TX FIFO empty elements above watermark IRQ occurred */
+	if (isr & (1 << 13)) {
+		can.base->IER &= ~(1 << 13);
+	}
+
+	return 1;
+}
+
+/**
+ * Initializes the CAN peripheral with the specified ID and baud rate.
+ * Configures the clock, MIO pins, and synchronization primitives.
+ */
+static int can_init(unsigned int can_id, int baudrate)
+{
+	uint32_t reg_val;
+
+	/* Find the matching time configuration */
+	bool timeCfgFound = false;
+	uint32_t timeCfgN = 0;
+	for (uint32_t i = 0; i < ARRAY_LENGTH(timeCfg); i++) {
+		if (timeCfg[i].baudrate_kbps == baudrate) {
+			timeCfgFound = true;
+			timeCfgN = i;
+		}
+	}
+	if (!timeCfgFound) {
+		debug("zynqmp-can: failed to configure desired baud rate\n");
+		return EXIT_FAILURE;
+	}
+
+	/* Initialize the clock */
+	if (init_clock(can_id) < 0) {
+		debug("zynqmp-can: cannot initialize clocks\n");
+		return EXIT_FAILURE;
+	}
+
+	/* Configure MIO pins */
+	if (config_mio_pin(info[can_id].rxPin) < 0 || config_mio_pin(info[can_id].txPin) < 0) {
+		debug("zynqmp-can: failed to initialize MIO pins\n");
+		return -EINVAL;
+	}
```

Comment:
Perhaps move it to a later position? There's a lot that can go wrong after this and there's no pins uninit anywhere